device-scripts/diag.sh

```diff
@@ -0,0 +1,186 @@
+#!/bin/sh
+#
+# Copyright (C) 2023 Foundries.io
+#
+# SPDX-License-Identifier: Apache-2.0
+#
+# Suggested setting with vim:
+# set ai ts=4 sw=4 et
+#
+# Authors:
+#    Tim Anderson <tim.anderson@foundries.io>
+#    Vanessa Maegima <vanessa.maegima@foundries.io>
+#    Caio Pereira <caio.pereira@foundries.io>
+
+VERSION=1.0
+
+echo "*** diag tool version ***"
+echo "Version: $VERSION"
+echo
+
+echo "*** os-release ***"
+cat /etc/os-release
+echo
+
+test -f /var/sota/current-target && echo -e "*** current_target ***\n$(cat /var/sota/current-target)"
+echo
+
+echo "*** active networks ***"
+for chk in $(ip addr | grep "^[0-9]" | sed 's/^[0-9]*://;s/:.*//')
+do
+    ifconfig $chk 2> /dev/null | grep -q inet
+    if [ $? -eq 0 ]
+     then
+         ifconfig $chk
+     fi
+done
+echo
+
+echo "*** Name servers ***"
+ls -l /etc/resolv.conf
+grep "^nameserver" /etc/resolv.conf
+echo
+
+echo "*** iptable configuration ***"
+    iptables -L -v
+echo
+
+echo "*** /var/sota content ***"
+CONTENT=$(ls -l /var/sota 2> /dev/null)
+echo "$CONTENT"
+echo
+
+bad=0
+if [ -f /var/sota/sota.toml ]
+then
+    CFG=/var/sota/sota.toml
+    SERV=$(grep -m 1 "^server" $CFG | \
+        sed 's|^.*= "https://\(.*\)"|\1|')
+    PIN=$(grep "^pass" $CFG | sed 's/^.*= "//;s/"$//')
+    MODULE=$(grep "^module" $CFG | sed 's/^.*= "//;s/"$//')
+    ROOT_CERT_DIR=$(grep "^tls_cacert_path" $CFG | sed 's/^.*= "//;s/"$//')
+    if grep -q 'pkey_source = "file"' /var/sota/sota.toml
+    then
+        for FILE in client.pem root.crt pkey.pem
+        do
+            if [ ! -f /var/sota/${FILE} ]
+            then
+                echo "### missing /var/sota/$FILE ###"
+                bad=1
+            fi
+        done
+        if [ $bad -eq 0 ]
+        then
+            echo -e "\n*** checking certs ***"
+           openssl s_client -cert /var/sota/client.pem \
+              -CAfile $ROOT_CERT_DIR \
+              -key /var/sota/pkey.pem \
+              -connect $SERV < /dev/null || echo "### Failed ###"
+        else
+            echo -e "\n*** no cert checking connectivity ***"
+            openssl s_client -connect $SERV < /dev/null || echo "### Failed ###"
+        fi
+    else
+        CLIENT_ID=$(grep "^tls_clientcert_id" $CFG | sed 's/^.*= "//;s/"$//')
+        PKEY_ID=$(grep "^tls_pkey_id" $CFG | sed 's/^.*= "//;s/"$//')
+        if [ "$ROOT_CERT_DIR" = "/var/sota/root.crt" ]
+        then
+            echo -e "\n*** checking HSM certs ***"
+        else
+            echo -e "\n*** checking HSM+EL2GO certs ***"
+        fi
+        cat > /var/tmp/engine.cnf <<EOF
+openssl_conf = oc
+[oc]
+engines = eng
+[eng]
+pkcs11 = p11
+[p11]
+engine_id = pkcs11
+dynamic_path = /usr/lib/engines-3/pkcs11.so
+MODULE_PATH = $MODULE
+init = 0
+PIN = $PIN
+EOF
+        pkcs11-tool --module $MODULE --login \
```

Comment:
Might be good to check if pkcs11-tool is available as we will at some point drop it from the default package list (now that lmp-device-register is using libpkcs11 directly).

Reply:
is there an alternative?

Reply:
Not really, the idea to remove the tool is that it ends up being way too powerful and easy for an user to make a mistake with it. But it can still be installed depending on the user, which is why I would just check if it is available first and skip the check if not.